### DISCUSSION TITLE
Issue with extending WalletClient for custom actions using TypeScript

### DISCUSSION THREAD

**Discussion by DannyDelott** [wevm/viem#1599] [Question]
I'm working on a feature to allow users to extend their wallet client with custom actions. However, I'm facing a type issue with `walletClient.writeContract`. Below is a minimal reproduction of the problem. It seems to be related to `GetChain<TChain, undefined>`, but I'm not certain what's missing. This usage aligns with how `walletClient.writeContract` is typically used.

I would greatly appreciate any insights or suggestions to resolve this.

## Minimal Reproduction Code:

```ts
import { Account, Address, Chain, Transport, WalletClient } from "viem";
const erc20Abi = [ { inputs: [ { internalType: "address", name: "destination", type: "address", }, { internalType: "uint256", name: "amount", type: "uint256", }, ], name: "mint", outputs: [], stateMutability: "nonpayable", type: "function", }, ] as const;

function exampleErc20WalletActions<
  TTransport extends Transport = Transport,
  TChain extends Chain | undefined = Chain | undefined,
  TAccount extends Account | undefined = Account | undefined,
>(erc20Address: Address) {
  return (client: WalletClient<TTransport, TChain, TAccount>) => ({
    mintTokens: async ({
      destination,
      amount,
    }: {
      destination: Address;
      amount: bigint;
    }) => {
      const [account] = await client.getAddresses();
      return client.writeContract({
        address: erc20Address,
        abi: erc20Abi,
        functionName: "mint",
        args: [destination, amount],
        account,
      });
    };
  });
}
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgQQMaogVwHYwDQoAmhUApgM7kEDCAFgIbDYEAqU925ksBA6vQBsBpGNQHBSuOAF84AMygQQcAEQA3CSBUBuAFAZO8UlFQAmAAzIARsDgBeOAG1dcRC9dwmYTDHIAuJ3cPN2DgphhjbEEWAE8wUgCVemIyShU8INCokATVQgoYJnpCiGx0zOCYONyklIpyctCZDKaECo9wyOjqxMxw0wBWADZGprhsmvoQLFxRpqr43v7huY9pFo8AXQ3XCcSQcNWsGG9fAMdtzPIYYtIAWR96G3EqxOxSsHoYp+FVhZq5DhUCUyht1rpNnB6OQ4AZrnpdIDsMDgKU4KQAB5TMDCACiJgs-CEIjQIPIAB53Cw2BwuNAjBiIthCDCaZxuPAHGy6TwqXRGNh0YzJCy4PymHAAD5wHD5ORMUiEexihgS6Wy0jy7CKjYsNAYHAMpmi-UzeDq5mahVKhymw1SmWWrU63QAPgAFMYzJY6pQAshfeQAJQhOBkGCYKCC92ocSSGABInCURx3Dk6nsdn01ji5hwPXoM2ukN2V1wd1tYIHXAsCAAa0k-ih5BiyPLldC+WuRRBOw8UzNfekAQ7wS7hSiIP9gb0TQHhoCNgA5uFZ2sS2XRx44fBHPRC4bIQ56AB3RjwWMSXAAOiXJMDFHdQbXwXDkcFl-j15PUGAEWopQwOwwIVu0rjJCQ9QBF6FgBpBlB9sETzANBBKWDYiEeEiKKlAAclMNTVjAqxIVAS5No4449qiebzrglxjFCB64Jh0jPpk0hrmxejSLoQA

**Comment by jxom** [comment]
Here you go!

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgQQMaogVwHYwDQoAmhUApgM7kEDCAFgIbDYEAqU925ksBA6vQBsBpGNQHBSuOAF84AMygQQcAEQA3CSBUBuAFChu8JAHMRyYmUoUCAdyjAYpahFztU+ODACeYUnF72js6u9O4ACvTsICKkUOQy8orKAOQapCAA9KEwwC7kybq6GJzwsagATAAMyABGwHAAvHAA2rpwiG3tcExgmDDkAFwtnV0do6NMjlDYgiw+pEMq9BYU5Cp4I+Mz0YuEFDkzOS7rm6PevovLJKsn4zIbdwinXZOxMwJzF6qYk+UArAA2W53ODbBaqeggLC4YF3c7glQ-XD-IEPcbSNHtAC6mNBkIRIEmsLgWBgvX6Q2aOM25Bg9EcAFk+vQ6uJvItsC4wPQvCzhMT4Ys5Dh3LlsMCMbosXB6PFirS9LphdhRS44KQAB6QsDCACiUAqlX4QjMqs4AB5Oiw2BwuNBShrHNhCPEbZxDI04G67TwrXRGNh1Y7JC64P6mHAAD5wHB7ORMUiET3hwPR2OkePYRNolhoDA4B1O0N56HwNPOjMJpNNEsFqMxiuZ7O6AB8AAoylVzNdKENu5ZyABKMZwMgwTDTOBt1DiSQwIbG4SiWe4c3W9ju+2sFOsWu4FvDhotqdPUaE3AsCAAa0kgxl5C8KpPzz2tKY9LFuPakNLuOkQ1PcZXwOD8XD7FZKD0O4fwLIY6mMSYoK6aRD2PQDRnleBmlCfNcGlJp6BsRh4FMGB+1WChpxXGBByQ0YxwnQM7AcJwXBgNwYCoiRcAIdDxiuAchk7aoIKoZ5vzqISDS7Oovy6ZUzQAOXxRZzxgYlRkiYw72aYD3yOZgZShAtqRBGV0F-Z5ZFlfxAlYkJwkifEpiHOiZCQlC9GkXQgA

**Reply by DannyDelott** [reply]
Amazing! Thanks 🙏

**Reply by gndelia** [reply]
I opened this playground and I still see an error 
![image](https://github.com/user-attachments/assets/5b60c7c4-2e5c-460e-831c-6c84cdcc31af)

I have a similar issue on my code, and I found this discussion. I also get an error when creating a custom action. I also see there's a `as WriteContractParameters` - can this problem be solved without casting?

**Reply by DannyDelott** [reply]
@jxom It looks like this no longer works in the latest version of viem.
